sled-storage/src/manager.rs

```diff
@@ -0,0 +1,969 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! The storage manager task
+
+use std::collections::HashSet;
+
+use crate::dataset::{DatasetError, DatasetName};
+use crate::disk::{Disk, DiskError, RawDisk};
+use crate::error::Error;
+use crate::resources::StorageResources;
+use camino::Utf8PathBuf;
+use illumos_utils::zfs::{Mountpoint, Zfs};
+use illumos_utils::zpool::ZpoolName;
+use key_manager::StorageKeyRequester;
+use omicron_common::disk::DiskIdentity;
+use sled_hardware::DiskVariant;
+use slog::{error, info, o, warn, Logger};
+use tokio::sync::{mpsc, oneshot, watch};
+use tokio::time::{interval, Duration, MissedTickBehavior};
+use uuid::Uuid;
+
+// The size of the mpsc bounded channel used to communicate
+// between the `StorageHandle` and `StorageManager`.
```

Comment:
In principle I like this idea, but in practice I'm not sure it will actually work. We'd want to set a relatively large bound to not spam the log (even 10 seconds is a long time here, but we'd probably want more like 1-10 minutes). However due to the small amount of work, it's likely that if the capacity ever does grow it would do so within milliseconds and we'd only be able to capture the size of the queue if we got really lucky in our timing.

But the suggestion to log the capacity is still a great idea. We really do want to be able to evaluate things like this. You just reminded me of how I did this in the past. At the cost of a little more work, I stored the capacity in a histogram. We could do that and dump the histogram to logs periodically. We could also periodically push it up to oximeter, and we can certainly make it available via OMDB. I'm going to do exactly this. I collaborated on HdrHistogram years ago and it's fast and perfect for this "local" use case. We can actually track the histogram in the task loop/manager itself, and not have to worry about locks at all.

Reply:
Lol, put this comment in the wrong spot. I already repeated commentary in a draft change I'll push up eventually.

Reply:
In principle I like this idea, but in practice I'm not sure it will actually work. We'd want to set a relatively large bound to not spam the log (even 10 seconds is a long time here, but we'd probably want more like 1-10 minutes). However due to the small amount of work, it's likely that if the capacity ever does grow it would do so within milliseconds and we'd only be able to capture the size of the queue if we got really lucky in our timing.

But the suggestion to log the capacity is still a great idea. We really do want to be able to evaluate things like this. You just reminded me of how I did this in the past. At the cost of a little more work, I stored the capacity in a histogram. We could do that and dump the histogram to logs periodically. We could also periodically push it up to oximeter, and we can certainly make it available via OMDB. I'm going to do exactly this. I collaborated on HdrHistogram years ago and it's fast and perfect for this "local" use case. We can actually track the histogram in the task loop/manager itself, and not have to worry about locks at all.

Reply:
I added a longer block comment here. I still need to add the histogram support before I go ahead and resolve this.

Reply:
I ended up adding logging for when the capacity of the channel grew or shrunk by 10 or more entries.

Reply:
Done in 3998708ef